<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.5"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>TreeNode: PhyloTree.TreeBuilding.DistanceMatrix Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="Doxy.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectlogo"><img alt="Logo" src="Logo.svg"/></td>
  <td id="projectalign">
   <div id="projectname">TreeNode<span id="projectnumber">&#160;1.3.1</span>
   </div>
   <div id="projectbrief">A library to manipulate phylogenetic trees</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.5 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(document).ready(function(){initNavTree('class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">PhyloTree.TreeBuilding.DistanceMatrix Class Reference</div></div>
</div><!--header-->
<div class="contents">

<p>Contains methods to compute distance matrices.   
 <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#details">More...</a></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-static-methods" name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:a62031438133fd36a48334188308ef183"><td class="memItemLeft" align="right" valign="top">static float[][]&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a62031438133fd36a48334188308ef183">BuildFromAlignment</a> (IReadOnlyList&lt; byte[]&gt; sequences, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a> evolutionModel=EvolutionModel.Kimura, int numCores=-1, Action&lt; double &gt; progressCallback=null)</td></tr>
<tr class="memdesc:a62031438133fd36a48334188308ef183"><td class="mdescLeft">&#160;</td><td class="mdescRight">Build a distance matrix from aligned DNA sequences stored as a T:byte[] array where each byte corresponds to three positions.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a62031438133fd36a48334188308ef183">More...</a><br /></td></tr>
<tr class="separator:a62031438133fd36a48334188308ef183"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a51c377271f7f8611c87141902b4cd9f5"><td class="memItemLeft" align="right" valign="top">static float[][]&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a51c377271f7f8611c87141902b4cd9f5">BuildFromAlignment</a> (IReadOnlyList&lt; ushort[]&gt; sequences, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a> evolutionModel=EvolutionModel.Kimura, int numCores=-1, Action&lt; double &gt; progressCallback=null)</td></tr>
<tr class="memdesc:a51c377271f7f8611c87141902b4cd9f5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Build a distance matrix from aligned protein sequences stored as a T:ushort[] array where each ushort corresponds to two positions.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a51c377271f7f8611c87141902b4cd9f5">More...</a><br /></td></tr>
<tr class="separator:a51c377271f7f8611c87141902b4cd9f5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acc1d43a4ecc8753a7e9f371a374c1322"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#acc1d43a4ecc8753a7e9f371a374c1322">BuildFromAlignment</a> (float[][] allocatedMatrix, IReadOnlyList&lt; byte[]&gt; sequences, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a> evolutionModel=EvolutionModel.Kimura, int numCores=-1, Action&lt; double &gt; progressCallback=null)</td></tr>
<tr class="memdesc:acc1d43a4ecc8753a7e9f371a374c1322"><td class="mdescLeft">&#160;</td><td class="mdescRight">Build a distance matrix from aligned DNA sequences stored as a T:byte[] array where each byte corresponds to three positions.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#acc1d43a4ecc8753a7e9f371a374c1322">More...</a><br /></td></tr>
<tr class="separator:acc1d43a4ecc8753a7e9f371a374c1322"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abb62e226bb213f6b738f224ad3f803b2"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#abb62e226bb213f6b738f224ad3f803b2">BuildFromAlignment</a> (float[][] allocatedMatrix, IReadOnlyList&lt; ushort[]&gt; sequences, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a> evolutionModel=EvolutionModel.Kimura, int numCores=0, Action&lt; double &gt; progressCallback=null)</td></tr>
<tr class="memdesc:abb62e226bb213f6b738f224ad3f803b2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Build a distance matrix from aligned protein sequences stored as a T:ushort[] array where each ushort corresponds to two positions.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#abb62e226bb213f6b738f224ad3f803b2">More...</a><br /></td></tr>
<tr class="separator:abb62e226bb213f6b738f224ad3f803b2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9cdc251c6ccaa99880f316084bf0e5f4"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a9cdc251c6ccaa99880f316084bf0e5f4">BuildFromAlignment</a> (float[][] allocatedMatrix, IReadOnlyList&lt; string &gt; sequences, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#a6bf7a7f7e46f96f70a3457ef3e86ba2a">AlignmentType</a> alignmentType=AlignmentType.Autodetect, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a> evolutionModel=EvolutionModel.Kimura, int numCores=0, Action&lt; double &gt; progressCallback=null)</td></tr>
<tr class="memdesc:a9cdc251c6ccaa99880f316084bf0e5f4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Build a distance matrix from aligned DNA or protein sequences.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a9cdc251c6ccaa99880f316084bf0e5f4">More...</a><br /></td></tr>
<tr class="separator:a9cdc251c6ccaa99880f316084bf0e5f4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a80e9bb040b5e3dd789c838ec80fe545a"><td class="memItemLeft" align="right" valign="top">static float[][]&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a80e9bb040b5e3dd789c838ec80fe545a">BuildFromAlignment</a> (IReadOnlyList&lt; string &gt; sequences, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#a6bf7a7f7e46f96f70a3457ef3e86ba2a">AlignmentType</a> alignmentType=AlignmentType.Autodetect, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a> evolutionModel=EvolutionModel.Kimura, int numCores=0, Action&lt; double &gt; progressCallback=null)</td></tr>
<tr class="memdesc:a80e9bb040b5e3dd789c838ec80fe545a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Build a distance matrix from aligned DNA or protein sequences.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a80e9bb040b5e3dd789c838ec80fe545a">More...</a><br /></td></tr>
<tr class="separator:a80e9bb040b5e3dd789c838ec80fe545a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6515032cd1fd22026906c59d68c5b8b8"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a6515032cd1fd22026906c59d68c5b8b8">BootstrapReplicateFromAlignment</a> (float[][] allocatedMatrix, IReadOnlyList&lt; string &gt; sequences, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#a6bf7a7f7e46f96f70a3457ef3e86ba2a">AlignmentType</a> alignmentType=AlignmentType.Autodetect, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a> evolutionModel=EvolutionModel.Kimura, int numCores=0, Action&lt; double &gt; progressCallback=null)</td></tr>
<tr class="memdesc:a6515032cd1fd22026906c59d68c5b8b8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Build a bootstrap replicate of a distance matrix from aligned DNA or protein sequences.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a6515032cd1fd22026906c59d68c5b8b8">More...</a><br /></td></tr>
<tr class="separator:a6515032cd1fd22026906c59d68c5b8b8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1e29e3b3a37a2ddae6b22c8b2c4febe1"><td class="memItemLeft" align="right" valign="top">static float[][]&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a1e29e3b3a37a2ddae6b22c8b2c4febe1">BootstrapReplicateFromAlignment</a> (IReadOnlyList&lt; string &gt; sequences, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#a6bf7a7f7e46f96f70a3457ef3e86ba2a">AlignmentType</a> alignmentType=AlignmentType.Autodetect, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a> evolutionModel=EvolutionModel.Kimura, int numCores=0, Action&lt; double &gt; progressCallback=null)</td></tr>
<tr class="memdesc:a1e29e3b3a37a2ddae6b22c8b2c4febe1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Build a bootstrap replicate of a distance matrix from aligned DNA or protein sequences.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a1e29e3b3a37a2ddae6b22c8b2c4febe1">More...</a><br /></td></tr>
<tr class="separator:a1e29e3b3a37a2ddae6b22c8b2c4febe1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2ca8ede0ce7121f2ddc961866a2374fa"><td class="memItemLeft" align="right" valign="top">static float[][]&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a2ca8ede0ce7121f2ddc961866a2374fa">BuildFromAlignment</a> (Dictionary&lt; string, string &gt; alignment, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#a6bf7a7f7e46f96f70a3457ef3e86ba2a">AlignmentType</a> alignmentType=AlignmentType.Autodetect, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a> evolutionModel=EvolutionModel.Kimura, int numCores=0, Action&lt; double &gt; progressCallback=null)</td></tr>
<tr class="memdesc:a2ca8ede0ce7121f2ddc961866a2374fa"><td class="mdescLeft">&#160;</td><td class="mdescRight">Build a distance matrix from aligned DNA or protein sequences.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a2ca8ede0ce7121f2ddc961866a2374fa">More...</a><br /></td></tr>
<tr class="separator:a2ca8ede0ce7121f2ddc961866a2374fa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab5ad9238008d83b811d018de85785635"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#ab5ad9238008d83b811d018de85785635">BuildFromAlignment</a> (float[][] allocatedMatrix, Dictionary&lt; string, string &gt; alignment, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#a6bf7a7f7e46f96f70a3457ef3e86ba2a">AlignmentType</a> alignmentType=AlignmentType.Autodetect, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a> evolutionModel=EvolutionModel.Kimura, int numCores=0, Action&lt; double &gt; progressCallback=null)</td></tr>
<tr class="memdesc:ab5ad9238008d83b811d018de85785635"><td class="mdescLeft">&#160;</td><td class="mdescRight">Build a distance matrix from aligned DNA or protein sequences.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#ab5ad9238008d83b811d018de85785635">More...</a><br /></td></tr>
<tr class="separator:ab5ad9238008d83b811d018de85785635"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae4ad2f735520ea4374951ca3f18973e5"><td class="memItemLeft" align="right" valign="top">static float[][]&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#ae4ad2f735520ea4374951ca3f18973e5">BootstrapReplicateFromAlignment</a> (Dictionary&lt; string, string &gt; alignment, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#a6bf7a7f7e46f96f70a3457ef3e86ba2a">AlignmentType</a> alignmentType=AlignmentType.Autodetect, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a> evolutionModel=EvolutionModel.Kimura, int numCores=0, Action&lt; double &gt; progressCallback=null)</td></tr>
<tr class="memdesc:ae4ad2f735520ea4374951ca3f18973e5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Build a bootstrap replicate of a distance matrix from aligned DNA or protein sequences.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#ae4ad2f735520ea4374951ca3f18973e5">More...</a><br /></td></tr>
<tr class="separator:ae4ad2f735520ea4374951ca3f18973e5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af383ae65d7e7f9e4d6f0b4ccf5c7f5f0"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#af383ae65d7e7f9e4d6f0b4ccf5c7f5f0">BootstrapReplicateFromAlignment</a> (float[][] allocatedMatrix, Dictionary&lt; string, string &gt; alignment, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#a6bf7a7f7e46f96f70a3457ef3e86ba2a">AlignmentType</a> alignmentType=AlignmentType.Autodetect, <a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a> evolutionModel=EvolutionModel.Kimura, int numCores=0, Action&lt; double &gt; progressCallback=null)</td></tr>
<tr class="memdesc:af383ae65d7e7f9e4d6f0b4ccf5c7f5f0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Build a bootstrap replicate of a distance matrix from aligned DNA or protein sequences.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#af383ae65d7e7f9e4d6f0b4ccf5c7f5f0">More...</a><br /></td></tr>
<tr class="separator:af383ae65d7e7f9e4d6f0b4ccf5c7f5f0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a50a296a27555e932740cf5b787634719"><td class="memItemLeft" align="right" valign="top">static IEnumerable&lt; byte[]&gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a50a296a27555e932740cf5b787634719">ConvertDNASequences</a> (IEnumerable&lt; string &gt; sequences)</td></tr>
<tr class="memdesc:a50a296a27555e932740cf5b787634719"><td class="mdescLeft">&#160;</td><td class="mdescRight">Convert DNA sequences into T:byte[] arrays in which each byte corresponds to 3 positions.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a50a296a27555e932740cf5b787634719">More...</a><br /></td></tr>
<tr class="separator:a50a296a27555e932740cf5b787634719"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1cc4f9977f1ae541d4cf9aec1aed36d0"><td class="memItemLeft" align="right" valign="top">static List&lt; byte[]&gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a1cc4f9977f1ae541d4cf9aec1aed36d0">ConvertDNASequences</a> (IReadOnlyList&lt; string &gt; sequences)</td></tr>
<tr class="memdesc:a1cc4f9977f1ae541d4cf9aec1aed36d0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Convert DNA sequences into T:byte[] arrays in which each byte corresponds to 3 positions.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a1cc4f9977f1ae541d4cf9aec1aed36d0">More...</a><br /></td></tr>
<tr class="separator:a1cc4f9977f1ae541d4cf9aec1aed36d0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a78ed3f050f157aaa6f91ac950acae8e1"><td class="memItemLeft" align="right" valign="top">static IEnumerable&lt; byte[]&gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a78ed3f050f157aaa6f91ac950acae8e1">BootstrapDNASequences</a> (IEnumerable&lt; string &gt; sequences)</td></tr>
<tr class="memdesc:a78ed3f050f157aaa6f91ac950acae8e1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Computes a bootstrap replicate of a DNA sequence alignment.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a78ed3f050f157aaa6f91ac950acae8e1">More...</a><br /></td></tr>
<tr class="separator:a78ed3f050f157aaa6f91ac950acae8e1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a76e7950bd660b0beae4d0b063606dbbc"><td class="memItemLeft" align="right" valign="top">static List&lt; byte[]&gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a76e7950bd660b0beae4d0b063606dbbc">BootstrapDNASequences</a> (IReadOnlyList&lt; string &gt; sequences)</td></tr>
<tr class="memdesc:a76e7950bd660b0beae4d0b063606dbbc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Computes a bootstrap replicate of a DNA sequence alignment.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a76e7950bd660b0beae4d0b063606dbbc">More...</a><br /></td></tr>
<tr class="separator:a76e7950bd660b0beae4d0b063606dbbc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab0548d1e35c5d777c0bcd67a56cbc209"><td class="memItemLeft" align="right" valign="top">static IEnumerable&lt; ushort[]&gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#ab0548d1e35c5d777c0bcd67a56cbc209">ConvertProteinSequences</a> (IEnumerable&lt; string &gt; sequences)</td></tr>
<tr class="memdesc:ab0548d1e35c5d777c0bcd67a56cbc209"><td class="mdescLeft">&#160;</td><td class="mdescRight">Convert protein sequences into T:ushort[] arrays in which each ushort corresponds to 2 positions.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#ab0548d1e35c5d777c0bcd67a56cbc209">More...</a><br /></td></tr>
<tr class="separator:ab0548d1e35c5d777c0bcd67a56cbc209"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1d6d72c113c5ff74e8a6754ebb44e82d"><td class="memItemLeft" align="right" valign="top">static List&lt; ushort[]&gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a1d6d72c113c5ff74e8a6754ebb44e82d">ConvertProteinSequences</a> (IReadOnlyList&lt; string &gt; sequences)</td></tr>
<tr class="memdesc:a1d6d72c113c5ff74e8a6754ebb44e82d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Convert protein sequences into T:ushort[] arrays in which each ushort corresponds to 2 positions.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a1d6d72c113c5ff74e8a6754ebb44e82d">More...</a><br /></td></tr>
<tr class="separator:a1d6d72c113c5ff74e8a6754ebb44e82d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1736bbac99f7630b38dbffc257b4a413"><td class="memItemLeft" align="right" valign="top">static IEnumerable&lt; ushort[]&gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a1736bbac99f7630b38dbffc257b4a413">BootstrapProteinSequences</a> (IEnumerable&lt; string &gt; sequences)</td></tr>
<tr class="memdesc:a1736bbac99f7630b38dbffc257b4a413"><td class="mdescLeft">&#160;</td><td class="mdescRight">Computes a bootstrap replicate of a protein sequence alignment.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a1736bbac99f7630b38dbffc257b4a413">More...</a><br /></td></tr>
<tr class="separator:a1736bbac99f7630b38dbffc257b4a413"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3c197b3383b2b9d23fedb0cff3b1c3d2"><td class="memItemLeft" align="right" valign="top">static List&lt; ushort[]&gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a3c197b3383b2b9d23fedb0cff3b1c3d2">BootstrapProteinSequences</a> (IReadOnlyList&lt; string &gt; sequences)</td></tr>
<tr class="memdesc:a3c197b3383b2b9d23fedb0cff3b1c3d2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Computes a bootstrap replicate of a protein sequence alignment.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a3c197b3383b2b9d23fedb0cff3b1c3d2">More...</a><br /></td></tr>
<tr class="separator:a3c197b3383b2b9d23fedb0cff3b1c3d2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae6fd1c9b683fc927352ec12f688fd1e7"><td class="memItemLeft" align="right" valign="top">static ushort[]&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#ae6fd1c9b683fc927352ec12f688fd1e7">ConvertProteinSequence</a> (string sequence)</td></tr>
<tr class="memdesc:ae6fd1c9b683fc927352ec12f688fd1e7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Converts a protein sequence stored as a string into a T:ushort[] array. Each ushort contains 2 amino acid positions. The allowed symbols are the usual 20 1-letter amino acid abbreviations, plus U for Sec, O for Pyl, B for Asn or Asp, Z for Gln or Glu, J for Ile or Leu, X for any amino acid, * for stop codons and - for gaps (uppercase and lowercase). All other characters are treated as gaps. If the sequence length is not a multiple of 2, it is padded with gaps.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#ae6fd1c9b683fc927352ec12f688fd1e7">More...</a><br /></td></tr>
<tr class="separator:ae6fd1c9b683fc927352ec12f688fd1e7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a08a21244c22fcbd18a2b9c88846fe1d8"><td class="memItemLeft" align="right" valign="top">static float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a08a21244c22fcbd18a2b9c88846fe1d8">CompareProteinSequencesBLOSUM62</a> (ushort[] sequence1, ushort[] sequence2, int selfScore1, int selfScore2)</td></tr>
<tr class="memdesc:a08a21244c22fcbd18a2b9c88846fe1d8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Compares two protein sequences using the Scoredist correction with the BLOSUM62 matrix.   <a href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html#a08a21244c22fcbd18a2b9c88846fe1d8">More...</a><br /></td></tr>
<tr class="separator:a08a21244c22fcbd18a2b9c88846fe1d8"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p >Contains methods to compute distance matrices.  </p>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00066">66</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a78ed3f050f157aaa6f91ac950acae8e1" name="a78ed3f050f157aaa6f91ac950acae8e1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a78ed3f050f157aaa6f91ac950acae8e1">&#9670;&#160;</a></span>BootstrapDNASequences() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static IEnumerable&lt; byte[]&gt; PhyloTree.TreeBuilding.DistanceMatrix.BootstrapDNASequences </td>
          <td>(</td>
          <td class="paramtype">IEnumerable&lt; string &gt;&#160;</td>
          <td class="paramname"><em>sequences</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Computes a bootstrap replicate of a DNA sequence alignment.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sequences</td><td>The aligned DNA sequences.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An T:IEnumerable&lt;byte[]&gt; that, when enumerated, will contain the bootstrapped sequences, converted into T:byte[] arrays where each byte corresponds to 3 positions.</dd></dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">ArgumentOutOfRangeException</td><td>Thrown if not all of the sequences have the same length.</td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00533">533</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="a76e7950bd660b0beae4d0b063606dbbc" name="a76e7950bd660b0beae4d0b063606dbbc"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a76e7950bd660b0beae4d0b063606dbbc">&#9670;&#160;</a></span>BootstrapDNASequences() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static List&lt; byte[]&gt; PhyloTree.TreeBuilding.DistanceMatrix.BootstrapDNASequences </td>
          <td>(</td>
          <td class="paramtype">IReadOnlyList&lt; string &gt;&#160;</td>
          <td class="paramname"><em>sequences</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Computes a bootstrap replicate of a DNA sequence alignment.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sequences</td><td>The aligned DNA sequences.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An T:List&lt;byte[]&gt; that contains the bootstrapped sequences, converted into T:byte[] arrays where each byte corresponds to 3 positions.</dd></dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">ArgumentOutOfRangeException</td><td>Thrown if not all of the sequences have the same length.</td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00562">562</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="a1736bbac99f7630b38dbffc257b4a413" name="a1736bbac99f7630b38dbffc257b4a413"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1736bbac99f7630b38dbffc257b4a413">&#9670;&#160;</a></span>BootstrapProteinSequences() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static IEnumerable&lt; ushort[]&gt; PhyloTree.TreeBuilding.DistanceMatrix.BootstrapProteinSequences </td>
          <td>(</td>
          <td class="paramtype">IEnumerable&lt; string &gt;&#160;</td>
          <td class="paramname"><em>sequences</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Computes a bootstrap replicate of a protein sequence alignment.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sequences</td><td>The aligned protein sequences.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An T:IEnumerable&lt;ushort[]&gt; that, when enumerated, will contain the bootstrapped sequences, converted into T:ushort[] arrays where each ushort corresponds to 2 positions.</dd></dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">ArgumentOutOfRangeException</td><td>Thrown if not all of the sequences have the same length.</td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00596">596</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="a3c197b3383b2b9d23fedb0cff3b1c3d2" name="a3c197b3383b2b9d23fedb0cff3b1c3d2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3c197b3383b2b9d23fedb0cff3b1c3d2">&#9670;&#160;</a></span>BootstrapProteinSequences() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static List&lt; ushort[]&gt; PhyloTree.TreeBuilding.DistanceMatrix.BootstrapProteinSequences </td>
          <td>(</td>
          <td class="paramtype">IReadOnlyList&lt; string &gt;&#160;</td>
          <td class="paramname"><em>sequences</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Computes a bootstrap replicate of a protein sequence alignment.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sequences</td><td>The aligned protein sequences.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A T:List&lt;ushort[]&gt; that contains the bootstrapped sequences, converted into T:ushort[] arrays where each ushort corresponds to 2 positions.</dd></dl>
<dl class="exception"><dt>Exceptions</dt><dd>
  <table class="exception">
    <tr><td class="paramname">ArgumentOutOfRangeException</td><td>Thrown if not all of the sequences have the same length.</td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00625">625</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="ae4ad2f735520ea4374951ca3f18973e5" name="ae4ad2f735520ea4374951ca3f18973e5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae4ad2f735520ea4374951ca3f18973e5">&#9670;&#160;</a></span>BootstrapReplicateFromAlignment() <span class="overload">[1/4]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static float[][] PhyloTree.TreeBuilding.DistanceMatrix.BootstrapReplicateFromAlignment </td>
          <td>(</td>
          <td class="paramtype">Dictionary&lt; string, string &gt;&#160;</td>
          <td class="paramname"><em>alignment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#a6bf7a7f7e46f96f70a3457ef3e86ba2a">AlignmentType</a>&#160;</td>
          <td class="paramname"><em>alignmentType</em> = <code>AlignmentType.Autodetect</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a>&#160;</td>
          <td class="paramname"><em>evolutionModel</em> = <code>EvolutionModel.Kimura</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numCores</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Action&lt; double &gt;&#160;</td>
          <td class="paramname"><em>progressCallback</em> = <code>null</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Build a bootstrap replicate of a distance matrix from aligned DNA or protein sequences.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alignment</td><td>The aligned sequences. These will be resampled randomly.</td></tr>
    <tr><td class="paramname">alignmentType</td><td>The type of sequences in the alignment.</td></tr>
    <tr><td class="paramname">evolutionModel</td><td>The evolutionary model to use to compute the distance matrix.</td></tr>
    <tr><td class="paramname">numCores</td><td>The maximum number of threads to use, or -1 to let the runtime decide.</td></tr>
    <tr><td class="paramname">progressCallback</td><td>A method used to report progress.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A T:float[][] jagged array containing the lower-triangular distance matrix.</dd></dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00462">462</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="af383ae65d7e7f9e4d6f0b4ccf5c7f5f0" name="af383ae65d7e7f9e4d6f0b4ccf5c7f5f0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af383ae65d7e7f9e4d6f0b4ccf5c7f5f0">&#9670;&#160;</a></span>BootstrapReplicateFromAlignment() <span class="overload">[2/4]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void PhyloTree.TreeBuilding.DistanceMatrix.BootstrapReplicateFromAlignment </td>
          <td>(</td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>allocatedMatrix</em>[][], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Dictionary&lt; string, string &gt;&#160;</td>
          <td class="paramname"><em>alignment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#a6bf7a7f7e46f96f70a3457ef3e86ba2a">AlignmentType</a>&#160;</td>
          <td class="paramname"><em>alignmentType</em> = <code>AlignmentType.Autodetect</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a>&#160;</td>
          <td class="paramname"><em>evolutionModel</em> = <code>EvolutionModel.Kimura</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numCores</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Action&lt; double &gt;&#160;</td>
          <td class="paramname"><em>progressCallback</em> = <code>null</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Build a bootstrap replicate of a distance matrix from aligned DNA or protein sequences.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">allocatedMatrix</td><td>A pre-allocated T:float[][] jagged array that will contain the lower-triangular distance matrix.</td></tr>
    <tr><td class="paramname">alignment</td><td>The aligned sequences. These will be resampled randomly.</td></tr>
    <tr><td class="paramname">alignmentType</td><td>The type of sequences in the alignment.</td></tr>
    <tr><td class="paramname">evolutionModel</td><td>The evolutionary model to use to compute the distance matrix.</td></tr>
    <tr><td class="paramname">numCores</td><td>The maximum number of threads to use, or -1 to let the runtime decide.</td></tr>
    <tr><td class="paramname">progressCallback</td><td>A method used to report progress.</td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00476">476</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="a6515032cd1fd22026906c59d68c5b8b8" name="a6515032cd1fd22026906c59d68c5b8b8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6515032cd1fd22026906c59d68c5b8b8">&#9670;&#160;</a></span>BootstrapReplicateFromAlignment() <span class="overload">[3/4]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void PhyloTree.TreeBuilding.DistanceMatrix.BootstrapReplicateFromAlignment </td>
          <td>(</td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>allocatedMatrix</em>[][], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">IReadOnlyList&lt; string &gt;&#160;</td>
          <td class="paramname"><em>sequences</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#a6bf7a7f7e46f96f70a3457ef3e86ba2a">AlignmentType</a>&#160;</td>
          <td class="paramname"><em>alignmentType</em> = <code>AlignmentType.Autodetect</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a>&#160;</td>
          <td class="paramname"><em>evolutionModel</em> = <code>EvolutionModel.Kimura</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numCores</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Action&lt; double &gt;&#160;</td>
          <td class="paramname"><em>progressCallback</em> = <code>null</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Build a bootstrap replicate of a distance matrix from aligned DNA or protein sequences.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">allocatedMatrix</td><td>A pre-allocated T:float[][] jagged array that will contain the lower-triangular distance matrix.</td></tr>
    <tr><td class="paramname">sequences</td><td>The aligned sequences. These will be resampled randomly.</td></tr>
    <tr><td class="paramname">alignmentType</td><td>The type of sequences in the alignment.</td></tr>
    <tr><td class="paramname">evolutionModel</td><td>The evolutionary model to use to compute the distance matrix.</td></tr>
    <tr><td class="paramname">numCores</td><td>The maximum number of threads to use, or -1 to let the runtime decide.</td></tr>
    <tr><td class="paramname">progressCallback</td><td>A method used to report progress.</td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00355">355</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="a1e29e3b3a37a2ddae6b22c8b2c4febe1" name="a1e29e3b3a37a2ddae6b22c8b2c4febe1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1e29e3b3a37a2ddae6b22c8b2c4febe1">&#9670;&#160;</a></span>BootstrapReplicateFromAlignment() <span class="overload">[4/4]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static float[][] PhyloTree.TreeBuilding.DistanceMatrix.BootstrapReplicateFromAlignment </td>
          <td>(</td>
          <td class="paramtype">IReadOnlyList&lt; string &gt;&#160;</td>
          <td class="paramname"><em>sequences</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#a6bf7a7f7e46f96f70a3457ef3e86ba2a">AlignmentType</a>&#160;</td>
          <td class="paramname"><em>alignmentType</em> = <code>AlignmentType.Autodetect</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a>&#160;</td>
          <td class="paramname"><em>evolutionModel</em> = <code>EvolutionModel.Kimura</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numCores</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Action&lt; double &gt;&#160;</td>
          <td class="paramname"><em>progressCallback</em> = <code>null</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Build a bootstrap replicate of a distance matrix from aligned DNA or protein sequences.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sequences</td><td>The aligned sequences. These will be resampled randomly.</td></tr>
    <tr><td class="paramname">alignmentType</td><td>The type of sequences in the alignment.</td></tr>
    <tr><td class="paramname">evolutionModel</td><td>The evolutionary model to use to compute the distance matrix.</td></tr>
    <tr><td class="paramname">numCores</td><td>The maximum number of threads to use, or -1 to let the runtime decide.</td></tr>
    <tr><td class="paramname">progressCallback</td><td>A method used to report progress.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A T:float[][] jagged array containing the lower-triangular distance matrix.</dd></dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00411">411</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="a2ca8ede0ce7121f2ddc961866a2374fa" name="a2ca8ede0ce7121f2ddc961866a2374fa"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2ca8ede0ce7121f2ddc961866a2374fa">&#9670;&#160;</a></span>BuildFromAlignment() <span class="overload">[1/8]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static float[][] PhyloTree.TreeBuilding.DistanceMatrix.BuildFromAlignment </td>
          <td>(</td>
          <td class="paramtype">Dictionary&lt; string, string &gt;&#160;</td>
          <td class="paramname"><em>alignment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#a6bf7a7f7e46f96f70a3457ef3e86ba2a">AlignmentType</a>&#160;</td>
          <td class="paramname"><em>alignmentType</em> = <code>AlignmentType.Autodetect</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a>&#160;</td>
          <td class="paramname"><em>evolutionModel</em> = <code>EvolutionModel.Kimura</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numCores</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Action&lt; double &gt;&#160;</td>
          <td class="paramname"><em>progressCallback</em> = <code>null</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Build a distance matrix from aligned DNA or protein sequences.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alignment</td><td>The aligned sequences.</td></tr>
    <tr><td class="paramname">alignmentType</td><td>The type of sequences in the alignment.</td></tr>
    <tr><td class="paramname">evolutionModel</td><td>The evolutionary model to use to compute the distance matrix.</td></tr>
    <tr><td class="paramname">numCores</td><td>The maximum number of threads to use, or -1 to let the runtime decide.</td></tr>
    <tr><td class="paramname">progressCallback</td><td>A method used to report progress.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A T:float[][] jagged array containing the lower-triangular distance matrix.</dd></dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00434">434</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="ab5ad9238008d83b811d018de85785635" name="ab5ad9238008d83b811d018de85785635"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab5ad9238008d83b811d018de85785635">&#9670;&#160;</a></span>BuildFromAlignment() <span class="overload">[2/8]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void PhyloTree.TreeBuilding.DistanceMatrix.BuildFromAlignment </td>
          <td>(</td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>allocatedMatrix</em>[][], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Dictionary&lt; string, string &gt;&#160;</td>
          <td class="paramname"><em>alignment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#a6bf7a7f7e46f96f70a3457ef3e86ba2a">AlignmentType</a>&#160;</td>
          <td class="paramname"><em>alignmentType</em> = <code>AlignmentType.Autodetect</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a>&#160;</td>
          <td class="paramname"><em>evolutionModel</em> = <code>EvolutionModel.Kimura</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numCores</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Action&lt; double &gt;&#160;</td>
          <td class="paramname"><em>progressCallback</em> = <code>null</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Build a distance matrix from aligned DNA or protein sequences.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">allocatedMatrix</td><td>A pre-allocated T:float[][] jagged array that will contain the lower-triangular distance matrix.</td></tr>
    <tr><td class="paramname">alignment</td><td>The aligned sequences.</td></tr>
    <tr><td class="paramname">alignmentType</td><td>The type of sequences in the alignment.</td></tr>
    <tr><td class="paramname">evolutionModel</td><td>The evolutionary model to use to compute the distance matrix.</td></tr>
    <tr><td class="paramname">numCores</td><td>The maximum number of threads to use, or -1 to let the runtime decide.</td></tr>
    <tr><td class="paramname">progressCallback</td><td>A method used to report progress.</td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00448">448</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="acc1d43a4ecc8753a7e9f371a374c1322" name="acc1d43a4ecc8753a7e9f371a374c1322"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acc1d43a4ecc8753a7e9f371a374c1322">&#9670;&#160;</a></span>BuildFromAlignment() <span class="overload">[3/8]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void PhyloTree.TreeBuilding.DistanceMatrix.BuildFromAlignment </td>
          <td>(</td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>allocatedMatrix</em>[][], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">IReadOnlyList&lt; byte[]&gt;&#160;</td>
          <td class="paramname"><em>sequences</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a>&#160;</td>
          <td class="paramname"><em>evolutionModel</em> = <code>EvolutionModel.Kimura</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numCores</em> = <code>-1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Action&lt; double &gt;&#160;</td>
          <td class="paramname"><em>progressCallback</em> = <code>null</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Build a distance matrix from aligned DNA sequences stored as a T:byte[] array where each byte corresponds to three positions.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">allocatedMatrix</td><td>A pre-allocated T:float[][] jagged array that will contain the lower-triangular distance matrix.</td></tr>
    <tr><td class="paramname">sequences</td><td>The aligned sequences.</td></tr>
    <tr><td class="paramname">evolutionModel</td><td>The evolutionary model to use to compute the distance matrix.</td></tr>
    <tr><td class="paramname">numCores</td><td>The maximum number of threads to use, or -1 to let the runtime decide.</td></tr>
    <tr><td class="paramname">progressCallback</td><td>A method used to report progress.</td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00171">171</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="a9cdc251c6ccaa99880f316084bf0e5f4" name="a9cdc251c6ccaa99880f316084bf0e5f4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9cdc251c6ccaa99880f316084bf0e5f4">&#9670;&#160;</a></span>BuildFromAlignment() <span class="overload">[4/8]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void PhyloTree.TreeBuilding.DistanceMatrix.BuildFromAlignment </td>
          <td>(</td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>allocatedMatrix</em>[][], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">IReadOnlyList&lt; string &gt;&#160;</td>
          <td class="paramname"><em>sequences</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#a6bf7a7f7e46f96f70a3457ef3e86ba2a">AlignmentType</a>&#160;</td>
          <td class="paramname"><em>alignmentType</em> = <code>AlignmentType.Autodetect</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a>&#160;</td>
          <td class="paramname"><em>evolutionModel</em> = <code>EvolutionModel.Kimura</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numCores</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Action&lt; double &gt;&#160;</td>
          <td class="paramname"><em>progressCallback</em> = <code>null</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Build a distance matrix from aligned DNA or protein sequences.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">allocatedMatrix</td><td>A pre-allocated T:float[][] jagged array that will contain the lower-triangular distance matrix.</td></tr>
    <tr><td class="paramname">sequences</td><td>The aligned sequences.</td></tr>
    <tr><td class="paramname">alignmentType</td><td>The type of sequences in the alignment.</td></tr>
    <tr><td class="paramname">evolutionModel</td><td>The evolutionary model to use to compute the distance matrix.</td></tr>
    <tr><td class="paramname">numCores</td><td>The maximum number of threads to use, or -1 to let the runtime decide.</td></tr>
    <tr><td class="paramname">progressCallback</td><td>A method used to report progress.</td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00252">252</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="abb62e226bb213f6b738f224ad3f803b2" name="abb62e226bb213f6b738f224ad3f803b2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abb62e226bb213f6b738f224ad3f803b2">&#9670;&#160;</a></span>BuildFromAlignment() <span class="overload">[5/8]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static void PhyloTree.TreeBuilding.DistanceMatrix.BuildFromAlignment </td>
          <td>(</td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>allocatedMatrix</em>[][], </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">IReadOnlyList&lt; ushort[]&gt;&#160;</td>
          <td class="paramname"><em>sequences</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a>&#160;</td>
          <td class="paramname"><em>evolutionModel</em> = <code>EvolutionModel.Kimura</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numCores</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Action&lt; double &gt;&#160;</td>
          <td class="paramname"><em>progressCallback</em> = <code>null</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Build a distance matrix from aligned protein sequences stored as a T:ushort[] array where each ushort corresponds to two positions.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">allocatedMatrix</td><td>A pre-allocated T:float[][] jagged array that will contain the lower-triangular distance matrix.</td></tr>
    <tr><td class="paramname">sequences</td><td>The aligned sequences.</td></tr>
    <tr><td class="paramname">evolutionModel</td><td>The evolutionary model to use to compute the distance matrix.</td></tr>
    <tr><td class="paramname">numCores</td><td>The maximum number of threads to use, or -1 to let the runtime decide.</td></tr>
    <tr><td class="paramname">progressCallback</td><td>A method used to report progress.</td></tr>
  </table>
  </dd>
</dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00211">211</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="a62031438133fd36a48334188308ef183" name="a62031438133fd36a48334188308ef183"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a62031438133fd36a48334188308ef183">&#9670;&#160;</a></span>BuildFromAlignment() <span class="overload">[6/8]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static float[][] PhyloTree.TreeBuilding.DistanceMatrix.BuildFromAlignment </td>
          <td>(</td>
          <td class="paramtype">IReadOnlyList&lt; byte[]&gt;&#160;</td>
          <td class="paramname"><em>sequences</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a>&#160;</td>
          <td class="paramname"><em>evolutionModel</em> = <code>EvolutionModel.Kimura</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numCores</em> = <code>-1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Action&lt; double &gt;&#160;</td>
          <td class="paramname"><em>progressCallback</em> = <code>null</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Build a distance matrix from aligned DNA sequences stored as a T:byte[] array where each byte corresponds to three positions.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sequences</td><td>The aligned sequences.</td></tr>
    <tr><td class="paramname">evolutionModel</td><td>The evolutionary model to use to compute the distance matrix.</td></tr>
    <tr><td class="paramname">numCores</td><td>The maximum number of threads to use, or -1 to let the runtime decide.</td></tr>
    <tr><td class="paramname">progressCallback</td><td>A method used to report progress.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A T:float[][] jagged array containing the lower-triangular distance matrix.</dd></dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00127">127</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="a80e9bb040b5e3dd789c838ec80fe545a" name="a80e9bb040b5e3dd789c838ec80fe545a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a80e9bb040b5e3dd789c838ec80fe545a">&#9670;&#160;</a></span>BuildFromAlignment() <span class="overload">[7/8]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static float[][] PhyloTree.TreeBuilding.DistanceMatrix.BuildFromAlignment </td>
          <td>(</td>
          <td class="paramtype">IReadOnlyList&lt; string &gt;&#160;</td>
          <td class="paramname"><em>sequences</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#a6bf7a7f7e46f96f70a3457ef3e86ba2a">AlignmentType</a>&#160;</td>
          <td class="paramname"><em>alignmentType</em> = <code>AlignmentType.Autodetect</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a>&#160;</td>
          <td class="paramname"><em>evolutionModel</em> = <code>EvolutionModel.Kimura</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numCores</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Action&lt; double &gt;&#160;</td>
          <td class="paramname"><em>progressCallback</em> = <code>null</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Build a distance matrix from aligned DNA or protein sequences.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sequences</td><td>The aligned sequences.</td></tr>
    <tr><td class="paramname">alignmentType</td><td>The type of sequences in the alignment.</td></tr>
    <tr><td class="paramname">evolutionModel</td><td>The evolutionary model to use to compute the distance matrix.</td></tr>
    <tr><td class="paramname">numCores</td><td>The maximum number of threads to use, or -1 to let the runtime decide.</td></tr>
    <tr><td class="paramname">progressCallback</td><td>A method used to report progress.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A T:float[][] jagged array containing the lower-triangular distance matrix.</dd></dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00332">332</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="a51c377271f7f8611c87141902b4cd9f5" name="a51c377271f7f8611c87141902b4cd9f5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a51c377271f7f8611c87141902b4cd9f5">&#9670;&#160;</a></span>BuildFromAlignment() <span class="overload">[8/8]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static float[][] PhyloTree.TreeBuilding.DistanceMatrix.BuildFromAlignment </td>
          <td>(</td>
          <td class="paramtype">IReadOnlyList&lt; ushort[]&gt;&#160;</td>
          <td class="paramname"><em>sequences</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html#ac115de283ef367572afde3b6a0c05785">EvolutionModel</a>&#160;</td>
          <td class="paramname"><em>evolutionModel</em> = <code>EvolutionModel.Kimura</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>numCores</em> = <code>-1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">Action&lt; double &gt;&#160;</td>
          <td class="paramname"><em>progressCallback</em> = <code>null</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Build a distance matrix from aligned protein sequences stored as a T:ushort[] array where each ushort corresponds to two positions.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sequences</td><td>The aligned sequences.</td></tr>
    <tr><td class="paramname">evolutionModel</td><td>The evolutionary model to use to compute the distance matrix.</td></tr>
    <tr><td class="paramname">numCores</td><td>The maximum number of threads to use, or -1 to let the runtime decide.</td></tr>
    <tr><td class="paramname">progressCallback</td><td>A method used to report progress.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A T:float[][] jagged array containing the lower-triangular distance matrix.</dd></dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00149">149</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="a08a21244c22fcbd18a2b9c88846fe1d8" name="a08a21244c22fcbd18a2b9c88846fe1d8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a08a21244c22fcbd18a2b9c88846fe1d8">&#9670;&#160;</a></span>CompareProteinSequencesBLOSUM62()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static float PhyloTree.TreeBuilding.DistanceMatrix.CompareProteinSequencesBLOSUM62 </td>
          <td>(</td>
          <td class="paramtype">ushort[]&#160;</td>
          <td class="paramname"><em>sequence1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">ushort[]&#160;</td>
          <td class="paramname"><em>sequence2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>selfScore1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>selfScore2</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Compares two protein sequences using the Scoredist correction with the BLOSUM62 matrix.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sequence1</td><td>The first sequence.</td></tr>
    <tr><td class="paramname">sequence2</td><td>The second sequence.</td></tr>
    <tr><td class="paramname">selfScore1</td><td>The score obtained by comparing <em>sequence1</em>  with itself.</td></tr>
    <tr><td class="paramname">selfScore2</td><td>The score obtained by comparing <em>sequence2</em>  with itself.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The distance between the two sequences.</dd></dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8_protein_8cs_source.html#l00779">779</a> of file <a class="el" href="_distance_matrix_8_protein_8cs_source.html">DistanceMatrix.Protein.cs</a>.</p>

</div>
</div>
<a id="a50a296a27555e932740cf5b787634719" name="a50a296a27555e932740cf5b787634719"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a50a296a27555e932740cf5b787634719">&#9670;&#160;</a></span>ConvertDNASequences() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static IEnumerable&lt; byte[]&gt; PhyloTree.TreeBuilding.DistanceMatrix.ConvertDNASequences </td>
          <td>(</td>
          <td class="paramtype">IEnumerable&lt; string &gt;&#160;</td>
          <td class="paramname"><em>sequences</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Convert DNA sequences into T:byte[] arrays in which each byte corresponds to 3 positions.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sequences</td><td>The sequences to convert.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An T:IEnumerable&lt;byte[]&gt; that, when enumerated, will contain the converted sequences.</dd></dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00486">486</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="a1cc4f9977f1ae541d4cf9aec1aed36d0" name="a1cc4f9977f1ae541d4cf9aec1aed36d0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1cc4f9977f1ae541d4cf9aec1aed36d0">&#9670;&#160;</a></span>ConvertDNASequences() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static List&lt; byte[]&gt; PhyloTree.TreeBuilding.DistanceMatrix.ConvertDNASequences </td>
          <td>(</td>
          <td class="paramtype">IReadOnlyList&lt; string &gt;&#160;</td>
          <td class="paramname"><em>sequences</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Convert DNA sequences into T:byte[] arrays in which each byte corresponds to 3 positions.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sequences</td><td>The sequences to convert.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A T:List&lt;byte[]&gt; that contains the converted sequences.</dd></dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00499">499</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="ae6fd1c9b683fc927352ec12f688fd1e7" name="ae6fd1c9b683fc927352ec12f688fd1e7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae6fd1c9b683fc927352ec12f688fd1e7">&#9670;&#160;</a></span>ConvertProteinSequence()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static ushort[] PhyloTree.TreeBuilding.DistanceMatrix.ConvertProteinSequence </td>
          <td>(</td>
          <td class="paramtype">string&#160;</td>
          <td class="paramname"><em>sequence</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Converts a protein sequence stored as a string into a T:ushort[] array. Each ushort contains 2 amino acid positions. The allowed symbols are the usual 20 1-letter amino acid abbreviations, plus U for Sec, O for Pyl, B for Asn or Asp, Z for Gln or Glu, J for Ile or Leu, X for any amino acid, * for stop codons and - for gaps (uppercase and lowercase). All other characters are treated as gaps. If the sequence length is not a multiple of 2, it is padded with gaps.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sequence</td><td>The sequence to convert.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A T:ushort[] array representing the sequence.</dd></dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8_protein_8cs_source.html#l00053">53</a> of file <a class="el" href="_distance_matrix_8_protein_8cs_source.html">DistanceMatrix.Protein.cs</a>.</p>

</div>
</div>
<a id="ab0548d1e35c5d777c0bcd67a56cbc209" name="ab0548d1e35c5d777c0bcd67a56cbc209"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab0548d1e35c5d777c0bcd67a56cbc209">&#9670;&#160;</a></span>ConvertProteinSequences() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static IEnumerable&lt; ushort[]&gt; PhyloTree.TreeBuilding.DistanceMatrix.ConvertProteinSequences </td>
          <td>(</td>
          <td class="paramtype">IEnumerable&lt; string &gt;&#160;</td>
          <td class="paramname"><em>sequences</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Convert protein sequences into T:ushort[] arrays in which each ushort corresponds to 2 positions.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sequences</td><td>The sequences to convert.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>An T:IEnumerable&lt;ushort[]&gt; that, when enumerated, will contain the converted sequences.</dd></dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00572">572</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<a id="a1d6d72c113c5ff74e8a6754ebb44e82d" name="a1d6d72c113c5ff74e8a6754ebb44e82d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1d6d72c113c5ff74e8a6754ebb44e82d">&#9670;&#160;</a></span>ConvertProteinSequences() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static List&lt; ushort[]&gt; PhyloTree.TreeBuilding.DistanceMatrix.ConvertProteinSequences </td>
          <td>(</td>
          <td class="paramtype">IReadOnlyList&lt; string &gt;&#160;</td>
          <td class="paramname"><em>sequences</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Convert protein sequences into T:ushort[] arrays in which each ushort corresponds to 2 positions.  </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sequences</td><td>The sequences to convert.</td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A T:List&lt;ushort[]&gt; that contains the converted sequences.</dd></dl>

<p class="definition">Definition at line <a class="el" href="_distance_matrix_8cs_source.html#l00585">585</a> of file <a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a>.</p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>TreeBuilding/<a class="el" href="_distance_matrix_8cs_source.html">DistanceMatrix.cs</a></li>
<li>TreeBuilding/<a class="el" href="_distance_matrix_8_protein_8cs_source.html">DistanceMatrix.Protein.cs</a></li>
<li>TreeBuilding/<a class="el" href="_distance_matrix_8_d_n_a_8cs_source.html">DistanceMatrix.DNA.cs</a></li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><a class="el" href="namespace_phylo_tree.html">PhyloTree</a></li><li class="navelem"><a class="el" href="namespace_phylo_tree_1_1_tree_building.html">TreeBuilding</a></li><li class="navelem"><a class="el" href="class_phylo_tree_1_1_tree_building_1_1_distance_matrix.html">DistanceMatrix</a></li>
    <li class="footer">Generated by <a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.5 </li>
  </ul>
</div>
</body>
</html>
